cmd/katapult/data_centers.go

```diff
@@ -27,21 +29,34 @@ func listDataCentersCmd(client dataCentersClient) *cobra.Command {
 				return err
 			}
 
-			for _, dc := range dcs {
-				_, _ = fmt.Fprintf(
-					cmd.OutOrStdout(),
-					" - %s (%s) [%s] / %s\n",
-					dc.Name, dc.Permalink, dc.ID, dc.Country.Name,
-				)
+			if strings.ToLower(cmd.Flag("output").Value.String()) == "json" {
+				j, err := json.Marshal(dcs)
+				if err != nil {
+					return err
+				}
+				_, _ = cmd.OutOrStdout().Write(append(j, '\n'))
+			} else {
+				for _, dc := range dcs {
+					_, _ = fmt.Fprintf(
+						cmd.OutOrStdout(),
+						" - %s (%s) [%s] / %s\n",
+						dc.Name, dc.Permalink, dc.ID, dc.Country.Name,
+					)
+				}
 			}
 
 			return nil
 		},
 	}
+
+	flags := cmd.PersistentFlags()
+	flags.StringP("output", "o", "text", "Defines the output type of the data centers. Can be text or json.")
```

Comment:
Perhaps we could make this a global flag ? Wonder what @jimeh thinks

Reply:
Hmm, yeah I think it probably should be. And I think a more global approach to how we render the different things might yield easier to maintain code in the long term, though it's a bit more work, so I'm not sure. I'll outline what I have in mind, and see what you all think :)

Proposal:

We define a set of output types, which represent the various types of things we want to return. For example a `DataCenterOutput` struct, and a `DataCentersListOutput` type of `[]*DataCenterOutput`.

Each command then populates output types and also defines a Go text/template to render its own output types to plain text. Then a common rendering function accepts the output, text template, and `--output` flag value, and will either marshal it to JSON, YAML, or render it via the template. This would also allow custom templates to be specified via the command line, similar to Docker's CLI.

Of course we don't have to define `*Output` types for everything, if the struct from go-katapult is suitable we could simply use that, with the caveat that the CLI tool's output is more tightly coupled to go-katapult and more likely to change over time.

Thoughts?

Reply:
I quite like this. Definitely seems quite appropriate, and lets us break up the testing of the commands nicely (e.g test the actual logic fetches the correct data and seperately test the serialisation to the user specififed format)

Reply:
We need to approach logging seperately as well, and ensure that it goes to STDERR or something so it won't pollute people trying to read the JSON output.